chore: instantiatable names pallet

ntn-x2 opened the pull request:
Part of https://github.com/KILTprotocol/ticket/issues/3650.

Makes the web3name pallet instantiatable. I did not rename the pallet/crate itself, and was not sure if I wanted to do it. That could break more things and I thought we might want to leave everything else as-is.

Ad96el reviewed (commented):
I think what you can do is define the default type for the instance in the web3names pallet, similar to how you did for the pallet struct. This would allow many of the changes to be dropped.

So for instance: pub type Web3NameOf<T, I> ->  pub type Web3NameOf<T, I =()>

Is there a reason, why you did not do that?

ntn-x2 commented:
@Ad96el I did that here: https://github.com/KILTprotocol/kilt-node/pull/779/commits/e239793f643791916998aa4fc97623c6088d6e45.

I did not think about it, but I did it anyway since this pallet is generic only in theory, but in practice it's very much bound to the rest of the KILT runtime. But ideally, the definition of a "canonical web3name" should live in the runtime common crate, where we would then add a second definition for a "dotname" in there. But for now I addressed your comment here. Let me know.

ntn-x2 commented on pallets/pallet-web3-names/src/mock.rs:
```diff
@@ -23,20 +23,20 @@ use crate::{
+	owner: Web3NameOwnerOf<T, I>,
+	name: Web3NameOf<T, I>,
 	block_number: BlockNumberFor<T>,
-	deposit: BalanceOf<T>,
+	deposit: BalanceOf<T, I>,
 ) {
-	CurrencyOf::<T>::hold(&HoldReason::Deposit.into(), &payer, deposit)
+	CurrencyOf::<T, I>::hold(&HoldReason::Deposit.into(), &payer, deposit)
 		.expect("Payer should have enough funds for deposit");
 
-	Names::<T>::insert(&owner, name.clone());
-	Owner::<T>::insert(
+	Names::<T, I>::insert(&owner, name.clone());
+	Owner::<T, I>::insert(
 		&name,
-		Web3OwnershipOf::<T> {
+		Web3OwnershipOf::<T, I> {
```
This is used also in benchmarking, so we need `I` to be generic.

Ad96el replied:
Oh I see

ntn-x2 commented on pallets/pallet-migration/src/lib.rs:
```diff
@@ -228,7 +228,7 @@ pub mod pallet {
 					let key_hash = <T as frame_system::Config>::Hashing::hash(&storage_key);
 					MigratedKeys::<T>::insert(key_hash, ());
-					pallet_web3_names::migrations::update_balance_for_w3n::<T>(key)
+					pallet_web3_names::migrations::update_balance_for_w3n::<T, _>(key)
```
It's not important actually, since we are only migrating our currently deployed instance, which has a `()` generic. So I removed the second generic from that function: https://github.com/KILTprotocol/kilt-node/pull/779/commits/d5703c9e0b9c41d824571958a4eb353ef29721f7.

ntn-x2 replied:
So, still good catch! 😄

Ad96el replied:
Yeah, I was confused about how it would work. Only the existing pallet needs the migration, but if a blanket is provided, how does the implementation know which instance should be executed. But now It's pretty clear - at least for me

Ad96el commented on pallets/pallet-web3-names/src/migrations.rs:
```diff
@@ -26,11 +26,10 @@ use crate::{AccountIdOf, Config, CurrencyOf, Error, HoldReason, Owner, Web3NameO
 {
 	let details = Owner::<T>::get(key).ok_or(Error::<T>::NotFound)?;
-	switch_reserved_to_hold::<AccountIdOf<T>, CurrencyOf<T>>(
+	switch_reserved_to_hold::<AccountIdOf<T>, CurrencyOf<T, _>>(
```
I think the blanket is not necessary here

Ad96el replied:
Okay Okay :ok_woman:

Ad96el reviewed (approved):
Last minor, not so important comment
